### PR TITLE
Run checks against pull requests

### LINKED ISSUE #21: Run native builds on push to PRs

### PR DESCRIPTION
**Pull request by lukechilds**
Now checks will run against every feature branch and PR push instead of just pushes to master.

Resolves #21

### CODE REVIEW

**Comment by lukechilds** [comment]
Builds are really slow due to the QEMU builds. Kind of annoying cos it'll just eat up all the job in the allowed concurrent job limit if we get a lot of pushes.

We can "skip" the QEMU jobs by with

```yaml
- name: Skip non native builds on non master pushes
  if: github.ref != 'master' && matrix.arch != 'amd64'
  run: exit 1
```

It resolves the issue of eating up all the concurrent jobs but technically that's a job fail so it pollutes check output with loads of "failed" jobs. Kinda defauts the purpose cos the overall status will show as failed.

**Comment by lukechilds** [comment]
Seems like it's just not possible with GitHub actions to conditionally skip a matrix job.

https://github.community/t5/GitHub-Actions/How-to-conditionally-include-exclude-items-in-matrix-eg-based-on/td-p/37871

It's shame cos you can do that on most other CI systems.

Looks like there was a "neutral" exit code in GitHub Action beta which would've allowed us to skip jobs. We could `exit 78` which would kill the job but it would be treated as a neutral kill and not count as a fail.

https://github.community/t5/GitHub-Actions/GitHub-Actions-quot-neutral-quot-exit-code-is-incorrectly/td-p/29051

However it was removed because some programs could actually `exit 78` in legitimate errors: https://twitter.com/ethomson/status/1163899559279497217

That tweet implies there will be a neutral exit code but I can't find any documentation on it so doesn't look possible for now.

**Comment by lukechilds** [comment]
TL;DR: Options are:

- Run entire build system on all pushes, QEMU builds will eat up jobs
- Run only native builds, QEMU builds show as failed on non-master pushes
- Copy/paste test.yml and remove QEMU builds from matrix and run on all branches, means maintaining two basically identical GitHub actions :/

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -31,6 +28,10 @@ jobs:
           - amd64
 
     steps:
+      - name: Skip non native builds on non master pushes
+        if: github.ref != 'master' && matrix.arch != 'amd64'
+        run: exit 1
```
Comment (meeDamian) [inline .github/workflows/test.yml]: That will cause the job to always be red on PRs, quite not ideal I'd say.

**Comment by meeDamian** [comment]
> Clone test.yml and remove QEMU builds from matrix and run on all branches, means maintaining two basically identical Actions

I think that's by far the best solution.

**Comment by meeDamian** [comment]
Also maybe we can go with sth like:

```yaml
on:
  pull_request:
    - '*'
```

And all arch-specific stuff can be removed too, making the workflow quite small, and nice.

**Comment by meeDamian** [comment]
Let's keep `test` as `test`, and add another one called `pull-request`?

**Comment by lukechilds** [comment]
So you don't wanna run tests on feature branches?

PRs only?

**Comment by lukechilds** [comment]
>And all arch-specific stuff can be removed too, making the workflow quite small, and nice.

Also I intentionally left the two files as similar as possible so the git diffs should be almost identical.

e.g I've left the `matrix.arch` in even though it only has one entry.

My reasoning was that this makes it much easier to maintain them both and keep them in sync, we can just diff them and see that there are only the minimal trigger/matrix changes we want.

Happy to clean up the PR version if you want but this might make maintaining them in parralallel going forward more difficult.

**Comment by meeDamian** [comment]
Without the arch stuff, the file becomes really, simple:

```yml
name: Build bitcoind on pull request

on:
  push:
    branches:
      - master

jobs:
  build:
    name: Build Bitcoind
    runs-on: ubuntu-18.04
    env:
      DOCKER_BUILDKIT: 1

    strategy:
      fail-fast: false
      matrix:
        subver:
          - '0.16'
          - '0.17'
          - '0.18'
          - '0.19'

    steps:
      - uses: actions/checkout@v1.0.0

      - name: Build Bitcoind
        run: docker build  -t bitcoind  ${{matrix.subver}}/

      - name: Print Bitcoind version
        run: |
          docker run --rm  --entrypoint=uname  bitcoind  -a
          docker run --rm bitcoind --version
```

And because of that simple to read might be better here than simple to compare?

**Comment by meeDamian** [comment]
> So you don't wanna run tests on feature branches?
> 
> PRs only?

We don't have any, will be thinking about it once we do :P

And even if we do, if there's a PR open for it to `master` it'll build.  So good 'nuff I reckon.

File: .github/workflows/pull-request.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+          - '0.16'
+          - '0.17'
+          - '0.18'
+          - '0.19'
+
+    steps:
+      - uses: actions/checkout@v1.0.0
+
+      - name: Build Bitcoind
+        if: matrix.subver != '0.17' || matrix.arch != 'arm32v7'
```
Comment (meeDamian) [inline .github/workflows/pull-request.yml]: delete

File: .github/workflows/pull-request.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        run: docker build  -t bitcoind  ${{matrix.subver}}/
+
+      - name: Print Bitcoind version
+        if: matrix.subver != '0.17' || matrix.arch != 'arm32v7'
```
Comment (meeDamian) [inline .github/workflows/pull-request.yml]: delete

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -0,0 +1,33 @@
+name: Build bitcoind on push
```
Comment (meeDamian) [inline .github/workflows/pull-request.yml]: > … on pull request

or 

> … on any PR

**Comment by meeDamian** [comment]
After you change workflow description, I'm happy with it :).

**Comment by lukechilds** [comment]
Fixed before you even commented 😛

Let's just wait for the builds to complete just to make sure I didn't bork anything.

**Comment by meeDamian** [comment]
Now I wonder why `qemu-perf` never runs 🤔

**Comment by lukechilds** [comment]
Does it not run on cron:

https://github.com/lncm/docker-bitcoind/blob/574252c8392167a21f2f8fbedfe31d2a37051544/.github/workflows/qemu-perf.yml#L5

https://github.com/lncm/docker-bitcoind/actions?query=workflow%3A%22Perf-check+qemu+versions%22

**Comment by meeDamian** [comment]
https://github.com/lncm/docker-bitcoind/pull/27 ^^

**Comment by lukechilds** [comment]
So all these builds were triggered manually?

https://github.com/lncm/docker-bitcoind/actions?query=workflow%3A%22Perf-check+qemu+versions%22

**Comment by meeDamian** [comment]
Lol, no - I'm just a bonehead 😂.
